src/services/job/update_sender_in_tx_message.ts

```diff
@@ -0,0 +1,140 @@
+/* eslint-disable no-await-in-loop */
+import { Service } from '@ourparentcenter/moleculer-decorators-extended';
+import { ServiceBroker } from 'moleculer';
+import { Transaction, BlockCheckpoint, TransactionMessage } from '../../models';
+import BullableService, { QueueHandler } from '../../base/bullable.service';
+import { BULL_JOB_NAME, SERVICE } from '../../common';
+import config from '../../../config.json' assert { type: 'json' };
+import knex from '../../common/utils/db_connection';
+
+@Service({
+  name: SERVICE.V1.JobService.UpdateSenderInTxMessages.key,
+  version: 1,
+})
+export default class UpdateSenderInTxMessages extends BullableService {
+  public constructor(public broker: ServiceBroker) {
+    super(broker);
+  }
+
+  @QueueHandler({
+    queueName: BULL_JOB_NAME.JOB_UPDATE_SENDER_IN_TX_MESSAGES,
+    jobName: BULL_JOB_NAME.JOB_UPDATE_SENDER_IN_TX_MESSAGES,
+  })
+  async updateSender(_payload: { lastBlockCrawled: number }) {
+    const blockCheckpoint = await BlockCheckpoint.query().findOne({
+      job_name: BULL_JOB_NAME.JOB_UPDATE_SENDER_IN_TX_MESSAGES,
+    });
+    this.logger.info(
+      `Update sender in transaction_message table start from block ${blockCheckpoint?.height}`
+    );
+    if (blockCheckpoint?.height === _payload.lastBlockCrawled) {
+      return;
+    }
+
+    let lastBlock =
+      (blockCheckpoint?.height ?? 0) +
+      config.jobUpdateSenderInTxMessages.blocksPerCall;
+    if (lastBlock > _payload.lastBlockCrawled) {
+      lastBlock = _payload.lastBlockCrawled;
+    }
+    await knex.transaction(async (trx) => {
+      const listTx = await Transaction.query()
+        .withGraphFetched('events.[attributes]')
+        .modifyGraph('events', (builder) => {
+          builder.orderBy('id', 'asc');
+        })
+        .modifyGraph('events.[attributes]', (builder) => {
+          builder.orderBy('index', 'asc');
+        })
+        .modifyGraph('messages', (builder) => {
+          builder.orderBy('id', 'asc');
+        })
+        .orderBy('id', 'asc')
+        .where('height', '>=', blockCheckpoint?.height ?? 0)
+        .andWhere('height', '<', lastBlock)
+        .transacting(trx);
+      await Promise.all(
+        listTx.map((tx) => {
+          const sender = this._findFirstAttribute(
+            tx.events,
+            'message',
+            'sender'
+          );
+          return TransactionMessage.query()
+            .patch({
+              sender,
+            })
+            .where({ tx_id: tx.id })
```

Comment:
mỗi tx update 1 lần là cũng lâu đấy, seren thì bao nhiêu tx ?

Reply:
487k tx anh ạ

Reply:
làm như này thử xem
https://stackoverflow.com/questions/18797608/update-multiple-rows-in-same-query-using-postgresql/18799497#18799497

Reply:
done